src/seqn/trans.lisp

```diff
@@ -0,0 +1,399 @@
+             (typep cadr 'vamp:constant))
+        (list (vamp:make-constant :const (+ (* (expt 2 (mcadr obj)) (vamp:const car))
+                               (vamp:const cadr))))
+        (list (vamp:make-infix :op :+
+                               :lhs (vamp:make-infix :op :*
+                                                     :lhs car
+                                                     :rhs (vamp:make-constant
+                                                           :const (expt 2 (mcadr obj))))
+                               :rhs cadr)))))
+
+(defmethod to-vampir ((obj seqn-decompose) inputs constraints)
+  (declare (ignore constraints))
+  (let* ((mcar (mcar obj))
+         (car (car inputs))
+         (rng  (vamp:make-application :func :range_n
+                                      :arguments (list (vamp:make-constant :const mcar)
+                                                       car)))
+         (lst  (list (vamp:make-constant :const (1- mcar))
+                     rng))
+         (dpth (range-depth rng)))
+    (if (typep car 'vamp:constant)
+        (if (>= (vamp:const car) (expt 2 (1- mcar)))
+            (list (vamp:make-constant :const 1)
+                  (vamp:make-constant :const (- (vamp:const car) (expt 2 (1- mcar)))))
+            (list (vamp:make-constant :const 0)
+                  car))
+        (if (= dpth 0)
+            (list (vamp:make-application :func :n_th
+                                         :arguments lst)
+                  (geb.vampir:combine
+                   (vamp:make-application :func :drop_ith
+                                          :arguments lst)))
+            (list (vamp:make-application :func :n_th
+                                         :arguments (list (vamp:make-constant :const (1- mcar))
+                                                          (geb.utils:apply-n dpth
+                                                                             (lambda (x)
+                                                                               (cadr
+                                                                                (vamp:arguments
+                                                                                 (car
+                                                                                  (vamp:arguments
+                                                                                   (cadr
+                                                                                    (vamp:arguments x)))))))
+                                                                             rng)))
+                  (geb.vampir:combine
+                   (vamp:make-application :func :drop_ith
+                                          :arguments lst)))))))
+
+(defun range-depth (x)
+  (let ((cadr (cadr (vamp:arguments x))))
+    (if (not (typep cadr 'vamp:application))
+        0
+        (1+ (range-depth (cadr (vamp:arguments (car (vamp:arguments cadr)))))))))
```

Comment:
you need to get rid of this argument pattern, make it have a reasonable name, as I can't make heads or tails of it.